Source/Scene/modelMaterialsCommon.js

```diff
@@ -0,0 +1,792 @@
+                            {
+                                type: WebGLConstants.FLOAT,
+                                value: spot.constantAttenuation
+                            };
+                            result[lightBaseName + 'LinearAttenuation'] =
+                            {
+                                type: WebGLConstants.FLOAT,
+                                value: spot.linearAttenuation
+                            };
+                            result[lightBaseName + 'QuadraticAttenuation'] =
+                            {
+                                type: WebGLConstants.FLOAT,
+                                value: spot.quadraticAttenuation
+                            };
+
+                            result[lightBaseName + 'FallOffAngle'] =
+                            {
+                                type: WebGLConstants.FLOAT,
+                                value: spot.fallOffAngle
+                            };
+                            result[lightBaseName + 'FallOffExponent'] =
+                            {
+                                type: WebGLConstants.FLOAT,
+                                value: spot.fallOffExponent
+                            };
+                        }
+                            break;
+                    }
+                    ++lightCount;
+                }
+            }
+        }
+
+        return result;
+    }
+
+    var techniqueCount = 0;
+    var shaderCount = 0;
+    var programCount = 0;
+    function generateTechnique(gltf, khrMaterialsCommon, attributes, lightParameters, jointCount) {
+        var techniques = gltf.techniques;
+        var shaders = gltf.shaders;
+        var programs = gltf.programs;
+        attributes = defaultValue(attributes, []);
+        var attributesCount = attributes.length;
+        var lightingModel = khrMaterialsCommon.technique.toUpperCase();
+        var lights;
+        if (defined(gltf.extensions) && defined(gltf.extensions.KHR_materials_common)) {
+            lights = gltf.extensions.KHR_materials_common.lights;
+        }
+        var hasSkinning = (jointCount > 0);
+        var values = khrMaterialsCommon.values;
+        var isDoubleSided = values.doubleSided;
+        delete values.doubleSided;
+
+        var vertexShader = 'precision highp float;\n';
+        var fragmentShader = 'precision highp float;\n';
+
+        // Generate IDs for our new objects
+        var techniqueId;
+        do {
+            techniqueId = 'technique' + (techniqueCount++).toString();
+        } while(defined(techniques[techniqueId]));
+
+        var shaderId;
+        var vertexShaderId;
+        var fragmentShaderId;
+        do {
+            shaderId = 'shader' + (shaderCount++).toString();
+            vertexShaderId = shaderId + 'VS';
+            fragmentShaderId = shaderId + 'FS';
+        } while(defined(shaders[vertexShaderId]) || defined(shaders[fragmentShaderId]));
```

Comment:
It's probably more elegant to separate this into two different `count` variables and two loops, right?  Could even create a helper for finding the next free id for a given dictionary object.